### ISSUE TITLE
IVsPackageInstaller2 interface install/update/downgrade/uninstall actions doesn't support batch operation

### ISSUE THREAD

**Issue by erdembayar** [NuGet/Home#12356]
Currently if I want to install/uninstall multiple packages via [IVsPackageInstaller2](https://github.com/NuGet/NuGet.Client/blob/ba3fa10c1ae0fc4b57f4fa01c8fda0abf47f865a/src/NuGet.Clients/NuGet.VisualStudio/Extensibility/IVsPackageInstaller2.cs) from another VS component then we have to sequentially call install/update/downgrade/uninstall operations 1 by 1 for each package inside each project.
In case we case `n` project and `m` packages in each then it takes `n*m` subsequent calls which is very slow, waste a lot of time with context switching. Also, some cases it simply fails due to NU1605 error because all involving packages need to be updated same time.

PMUI can already do install/update/downgrade/uninstall actions in batch, ideally, we need something like following:

```
public interface IVsPackageInstaller3
{
    // Project level packageIds
    void InstallPackages(List<string> sources, Project project, List<PackageIdentity> packageIds, bool ignoreDependencies);
   // Project level version ranges
    void InstallPackages(List<string> sources, Project project, List<PackageDependency> packageIds, bool ignoreDependencies);
   // Solution level packageIds
    void InstallPackages(List<string> sources, Solution solution,  List<(Project project, List<PackageIdentity>packageIds, bool removeDependencies>));
   // Solution level version ranges
    void InstallPackages(List<string> sources, Solution solution,  List<(Project project, List<PackageDependency>packageIds, bool removeDependencies>));
}
```

```
public interface IVsPackageUninstaller
{
    void UninstallPackages(Project project, List<string>packageIds, bool removeDependencies);
    void UninstallPackages(Solution solution, List<(Project project, List<string>packageIds, bool removeDependencies>));
}
```

Or more generic [ProjectActions](https://github.com/NuGet/NuGet.Client/blob/4a6b7bbad3fb31eb6d8c5a4242649419dedd0f29/src/NuGet.Clients/NuGet.VisualStudio.Internal.Contracts/ProjectAction.cs) would be great like PMUI
```
public interface IVsPackageActions
{
    // Project level actions
    void ExecuteActions(List<string> sources, Project project, List<ProjectAction> actions);
   // Solution level actions
    void InstallPackages(List<string> sources, Solution solution,  List<ProjectAction> actions);
}
```

PMUI execution flow: [Start](https://github.com/NuGet/NuGet.Client/blob/4a6b7bbad3fb31eb6d8c5a4242649419dedd0f29/src/NuGet.Clients/NuGet.PackageManagement.UI/Actions/UIActionEngine.cs#L461) On UI -> [Actual Execution](https://github.com/NuGet/NuGet.Client/blob/4a6b7bbad3fb31eb6d8c5a4242649419dedd0f29/src/NuGet.Clients/NuGet.PackageManagement.VisualStudio/Services/NuGetProjectManagerService.cs#L353) ->[Project level batch install/update/uninstall action executions](https://github.com/NuGet/NuGet.Client/blob/4a6b7bbad3fb31eb6d8c5a4242649419dedd0f29/src/NuGet.Core/NuGet.PackageManagement/NuGetPackageManager.cs#L2421).

Related issue: https://github.com/NuGet/Home/issues/12355

**Comment by erdembayar** [comment]
Is it possible to use nugetPackageManager.ExecuteNuGetProjectActionAsync from another VS component? Not sure how to get instance of NugetPackageManager.

**Comment by jeffkl** [comment]
Team Triage: Following a response from Andy, close the issues if we are going to use a different solution.

**Comment by erdembayar** [comment]
> Team Triage: Following a response from Andy, close the issues if we are going to use a different solution.

Closing based on today's meeting. We decided not to implement because we don't have clear signal from customers for adoption/usefulness of NuGetSolver, if data suggests wide adoption/usefulness, then we'll implement it.
Instead, we're going to modify csproj directly, it'll unblock another problem [Suggest and install version range if original package version is version range](https://devdiv.visualstudio.com/DevDiv/_workitems/edit/1713086).
